### PR TITLE
ENH: Generalized linear mixed models (Bayesian)

### PR DESCRIPTION
**Pull request by kshedden**
This is a basic implementation of Bayesian mixed GLM (GLIMMIX).  It is an alternative to using maximum likelihood that is somewhat easier to implement.  Laplace approximations should work for any GLM, but the more sophisticated variational Bayes approach is only implemented here for logistic models right now.  The VB approach is slow (5-10 minutes to fit a fairly complex model), but ML would also be slow for this problem.  
  
Update: it's not particularly slow anymore, due to improvements in the implementation.

### CODE REVIEW

**Comment by josef-pkt** [comment]
This pull request introduces 2 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-033b24bb8bc8b6484639a86029e07f415b98fbb3)

**new alerts:**

* 1 for Missing call to __init__ during object initialization
* 1 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by josef-pkt** [comment]
travis test failures are not small.

I don't think I will have time to review this anytime soon except for spot checking or very superficial review.
I think we can merge  this when you think it's ready.
However, I would prefer names that indicate that this is not one of our standard type of (frequentist) models, e.g. results does not inherit likelihood and wald inference. For example add a leading "B" to the name, "BMixedGLM".  I think we should reserve the name MixedGLM for a frequentist version, even if it is numerically essentially the same but has our standard results.

**Comment by josef-pkt** [comment]
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-9cadc92c942c4066569cf083a7c2d4aa00eb7088)

**new alerts:**

* 1 for Redundant assignment

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by josef-pkt** [comment]
This pull request introduces 3 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-2756f1faf33f685f6df8ce004911f47d31f59d34)

**new alerts:**

* 2 for Signature mismatch in overriding method
* 1 for Mismatch between signature and use of an overridden method

---

*Comment posted by [lgtm.com](https://lgtm.com)*

File: statsmodels/regression/bayes_mixed_glm.py [hunk 1]
```diff
@@ -0,0 +1,802 @@
...
+
+        summ = summary2.Summary()
+        summ.add_title(self.model.family.__class__.__name__ +
+                       " Mixed GLM Results")
+        summ.add_df(df)
+
+        return summ
+
+
+class BinomialBayesMixedGLM(_VariationalBayesMixedGLM, _BayesMixedGLM):
```
Comment (josef-pkt) [inline statsmodels/regression/bayes_mixed_glm.py]: currently _VariationalBayesMixedGLM already subclasses _BayesMixedGLM and is not an extra Mixin, just regular inheritance

Comment (kshedden) [reply statsmodels/regression/bayes_mixed_glm.py]: Thanks, fixed

**Comment by kshedden** [comment]
I will merge soon if no objections...

**Comment by josef-pkt** [comment]
I will skim it tonight or tomorrow

**Comment by josef-pkt** [comment]
general question I had:
Are there references for this, for the general method? The only reference I saw is the variational Bayes method article. It would be good to have some background reference in the docstrings or documentation.

Is there an analogous method in another package that can be used as background reference or as start to search for references?

File: statsmodels/regression/tests/test_bayes_mixed_glm.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    assert_allclose(rslt1.cov_params.flat[0:5], np.r_[
+        0.00790914, 0.00080666, -0.00050719, 0.00022648, 0.00046235],
+                    rtol=1e-4, atol=1e-4)
+
+    assert_allclose(rslt2.params[0:5], np.r_[
+        -0.07088814, -0.06373107, -0.22770786,  1.12923746, -1.26161339],
+                    rtol=1e-4, atol=1e-4)
+
+    assert_allclose(rslt2.cov_params[0:5], np.r_[
+        0.00747782, 0.0092554, 0.04508904, 0.02934488, 0.20312746],
```
Comment (josef-pkt) [inline statsmodels/regression/tests/test_bayes_mixed_glm.py]: where do the reference numbers come from?
regression test or from other package?

**Comment by josef-pkt** [comment]
I just realized when I wanted to look at it in my local checkout of this:
Why is this in `regression` and not in `genmod`?

File: statsmodels/regression/tests/test_bayes_mixed_glm.py (hunk 2 above)
Comment (kshedden) [reply statsmodels/regression/tests/test_bayes_mixed_glm.py]: It's regression testing.  I'm not aware of any other implementation of this method for GLIMMIX models.

**Comment by josef-pkt** [comment]
I was thinking about putting it in discrete, genmod or base.
If it requires a special class for each family, then discrete would be the location, if it is for general/several families (with possibly different link functions), then it would be genmod. If it is the base for a new estimation method, then it would be `base`.

I'm thinking of genmod for general linear exponential family, not for a specific estimation method (e.g. GEE versus Mixed). I think it's better to stick with subpackages by topic or model category instead of estimation method. 
(I don't think I will be a fan of a specific `bayes` directory unless we would need it to park a lot of tools for it.)
GLMixed or MixedGLM was targeted for genmod.

related: generalized method of moments GMM is currently in sandbox.regression. I was thinking that the base estimators will move to `base` and application for specific model categories into their corresponding subdirectory.

**Comment by josef-pkt** [comment]
a quick try https://gist.github.com/josef-pkt/9e295e68727fcaf5d95f6b0bb4284f8a
includes a Gaussian family excercise, produces results close to OLS with small shrinkage of random effects compared to OLS group fixed effects

glmm3 = _BayesMixedGLM(y, exog_fe, exog_vc, ident, families.Gaussian(), vcp_p=0.5)
rslt3 = glmm3.fit_map()

I'm stopping here, I didn't see anything major and getting through the details would take more time.

IMO it belongs into `genmod`, given that gaussian example works out of the box using the private superclass.

I think it can be merged after moving it, but there will still be some refactorings and improvements left for the next round, e.g. top part of summary. Some inference on fixed effects, ... (maybe inherit from Model)

**Comment by kshedden** [comment]
I've moved it to genmod and will merge after I review it one more time later today.

**Comment by josef-pkt** [comment]
needs to be added to the relevant api.py and to a .rst file in docs
plus a paragraph or three for the release notes

**Comment by kshedden** [comment]
I'm going to go ahead and merge this now.  I will add support for scale parameters in a separate PR.

**Pull request merged**